Modules/gcmodule.c

```diff
@@ -610,6 +610,19 @@ move_legacy_finalizers(PyGC_Head *unreachable, PyGC_Head *finalizers)
     }
 }
 
+static inline void
+restore_unreachable_mask(PyGC_Head *unreachable)
+{
+    PyGC_Head *gc, *next;
+    unreachable->_gc_next &= ~NEXT_MASK_UNREACHABLE;
```

Comment:
I believe list heads never have these goofy 😉 flags set.  So better to assert that the flag isn't set here.

Reply:
I added a call to  `validate_list` after `unreachable` returns to be a normal list and an assert at the begging for the list head. Notice that we can not add such a test to `unreachable->_gc_next` because `unreachable->_gc_next` will not have the bit set of the list is empty IIRC.

Reply:
Not sure I followed that.  `unreachable` (any gc list) is empty if and only if  `unreachable->gc_next == unreachable` .  No flags in the header pointers will ever be set regardless,  empty or not.  Indeed, that's another thing `validate_list` could verify.

Reply:
I may be missing something.... If you are referring to add:

```diff
assert((unreachable->_gc_next & NEXT_MASK_UNREACHABLE) == 0); 
```

that will fail as the `_gc_next` field has indeed the flag set:

```
❯ ./python -m test test_gc 
0:00:00 load avg: 1.22 Run tests sequentially
0:00:00 load avg: 1.22 [1/1] test_gc
python: Modules/gcmodule.c:620: clear_unreachable_mask: Assertion `(unreachable->_gc_next & NEXT_MASK_UNREACHABLE) == 0' failed.
Fatal Python error: Aborted
```

Or are you referring to the head itself?

```
    assert(((uintptr_t)unreachable & NEXT_MASK_UNREACHABLE) == 0);
    assert(((uintptr_t)unreachable & PREV_MASK_COLLECTING) == 0)
```

Apologies for the confusion!

Reply:
Nope , my apologies - the "rule" that a list head's `next` and `prev` pointers are never polluted with flags appears to be deliberately violated by `move_unreachable` for the `unreachable` list head's `next` pointer.   So your original code was both correct & necessary.